crates/ol/stf/src/assembly.rs

```diff
@@ -0,0 +1,266 @@
+        self.post_state_roots.header_state_root()
+    }
+
+    pub fn logs(&self) -> &[OLLog] {
+        &self.logs
+    }
+
+    /// Computes the block's logs root from the log.
+    pub fn compute_block_logs_root(&self) -> Buf32 {
+        // This is just a simple binary merkle tree.
+        if self.logs.is_empty() {
+            // Empty tree has null root
+            return Buf32::zero();
+        }
+
+        // Hash each log entry to create leaf nodes
+        let mut leaf_hashes: Vec<[u8; 32]> = self
+            .logs
+            .iter()
+            .map(|log| log.compute_hash_commitment().0)
+            .collect();
+
+        // BinaryMerkleTree requires power of two leaves, so pad if necessary
+        let next_power_of_two = leaf_hashes.len().next_power_of_two();
+        while leaf_hashes.len() < next_power_of_two {
+            // Pad with zero hashes
+            leaf_hashes.push([0u8; 32]);
+        }
+
+        // Build the merkle tree using Sha256Hasher
+        let tree = BinaryMerkleTree::from_leaves::<Sha256Hasher>(leaf_hashes)
+            .expect("power of two leaves should always succeed");
+
+        Buf32(*tree.root())
+    }
+}
+
+/// Performs execution using parts of a block on top of a state, producing
+/// records of its output that we can use to complete a header for that drafted
+/// block.
+///
+/// This closely aligns with `verify_block_classically`.
+pub fn execute_block_inputs<S: StateAccessor>(
+    state: &mut S,
+    block_context: BlockContext<'_>,
+    block_exec_input: BlockExecInput<'_>,
+) -> ExecResult<BlockExecOutputs> {
+    // 0. Construct the block exec context for tracking verification state
+    // across phases.
+    let output = ExecOutputBuffer::new_empty();
+
+    // 1. If it's the first block of the epoch, call process_epoch_initial.
+    if block_context.is_probably_epoch_initial() {
+        let init_ctx = block_context.to_epoch_initial_context();
+        chain_processing::process_epoch_initial(state, &init_ctx)?;
+    }
+
+    // 2. Process the slot start for every block.
+    chain_processing::process_slot_start(state, &block_context)?;
+
+    // 3. Call process_block_tx_segment for every block as usual.
+    let basic_ctx = BasicExecContext::new(*block_context.block_info(), &output);
+    let tx_ctx = TxExecContext::new(&basic_ctx, block_context.parent_header());
+    transaction_processing::process_block_tx_segment(
+        state,
+        block_exec_input.tx_segment(),
+        &tx_ctx,
+    )?;
+
+    // 4. Compute the state root and remember it.
+    let pre_manifest_state_root = state.compute_state_root()?;
+
+    // 5. If it's the last block of an epoch, then call process_block_manifests,
+    // and compute the final state root and remember it.
+    //
+    // Then we use this to figure out what our state commitments should be.
+    let post_state_roots = if let Some(manifest_container) = block_exec_input.manifest_container() {
```

Comment:
we have `is_probably_terminal()` which we don't seem to use

Reply:
I've made the terminal flagging more systematic/explicit and changed this to `is_body_terminal`.